pkg/build/prometheus/metrics_via_collectors.go

```diff
@@ -0,0 +1,159 @@
+	StageDurationBucketQuery = BuildSubsystem + separator + StageDuration + separator + "bucket"
+	StageDurationCountQuery  = BuildSubsystem + separator + StageDuration + separator + "count"
+	StageDurationSumQuery    = BuildSubsystem + separator + StageDuration + separator + "sum"
+	StepDuration             = "steps_duration_seconds"
+	StepDurationBucketQuery  = BuildSubsystem + separator + StepDuration + separator + "bucket"
+	StepDurationCountQuery   = BuildSubsystem + separator + StepDuration + separator + "count"
+	StepDurationSumQuery     = BuildSubsystem + separator + StepDuration + separator + "sum"
+)
+
+var (
+	// decided not to have a separate counter for failed builds, which have reasons,
+	// vs. the other "finished" builds phases, where the reason is not set
+	descTerminalBuildCount = prometheus.NewDesc(
+		BuildSubsystem+separator+TerminalBuildCount,
+		"Counts total teriminal builds by namespace, name, phase, and in case of failures, reason",
+		[]string{"phase", "reason", "namespace", "name"},
+		nil,
+	)
+	descActiveBuildCount = prometheus.NewDesc(
+		BuildSubsystem+separator+ActiveBuildCount,
+		"Show current counts of active builds by namespace, name, phase, and with start time noted",
+		[]string{"phase", "namespace", "name", "startime"},
+		nil,
+	)
+	descBuildDuration = prometheus.NewDesc(
+		BuildSubsystem+separator+BuildDuration,
+		"Histogram of durations in seconds of terminated builds, by termination type",
+		[]string{"phase", "namespace", "name"},
+		nil,
+	)
+	descStageDuration = prometheus.NewDesc(
+		BuildSubsystem+separator+StageDuration,
+		"Histogram of durations in seconds of stages that have finished, by stage name",
+		[]string{"stage", "namespace", "name"},
+		nil,
+	)
+	descStepDuration = prometheus.NewDesc(
+		BuildSubsystem+separator+StepDuration,
+		"Histogram of durations in seconds of steps that have finished, by stage name and step name",
+		[]string{"stage", "step", "namespace", "name"},
+		nil,
+	)
+	bc         = buildCollector{}
+	registered = false
+)
+
+type buildCollector struct {
+	lister internalversion.BuildLister
+}
+
+// kube-state-metrics leverages prometheus.Registerer but does not exist in the version of prometheus
+// vendored into origin
+
+func IntializeMetricsCollector(buildLister internalversion.BuildLister) {
+	bc.lister = buildLister
+	// unit tests unearthed duplicate registration with prometheus via multiple calls to new build controller
+	if !registered {
+		prometheus.MustRegister(&bc)
+		registered = true
+	}
+	glog.V(4).Info("build metrics registered with prometheus")
+}
+
+// Describe implements the prometheus.Collector interface.
+func (bc *buildCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- descTerminalBuildCount
+	ch <- descActiveBuildCount
+	ch <- descBuildDuration
+	ch <- descStageDuration
+	ch <- descStepDuration
+}
+
+// Collect implements the prometheus.Collector interface.
+func (bc *buildCollector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
OK, I can more closely mimic the method structure if that is deemed important.

On "reset", perhaps a) the comment needs to be better, or perhaps b) we are not entirely connected on this aspect yet.
per a), the "reset" is the decrement of gauge for a given build's active state.  Since the gauge is per build, we have to achieve a flow of 
- increment new
- decrement new, increment pending
- decrement pending, increment running
- decrement running, then bump the counter for the terminal states

The simpler form to me seemed to decrement or "reset" always, the increment as the specific state is detected.  So yes, there is a decrement/increment churn if a build stays in the same state, but the value will be correct when the method exits.

And it solves the "state machine" conundrum.  Apologies for not thinking of a better name.  Hopefully a specific example will properly illustrate.

What happens if the collection calls are too far apart, and after incrementing the gauge for new, no collection calls come between build going from new to pending to running, missing the transition to pending.  Where do we put the decrement of the new gauge ... can't assume it should always be in the block when we increment pending.

As to why a gauge is not used, 3 things
1) ahhh ... the "active" metrics were gauges :-) originally ... when I switched to this collector form, did a little too much copying pasting ... minimally I'll change those, thanks.
2) for the "terminal" builds metrics, I originally had those as counters since they were only incrementing (i.e. not decrementing as builds were pruned);  given the exchange with @bparees I assume we are OK with that .... while counters are sufficient, I'm fine with changing to gauges if that is what we want
3) for the time duration, counters / gauges are wrong ... assuming you are in agreement with histograms.  If not, please explain why